PCI: vmd: add PCH rootbus

sdurawa opened the pull request:
Introducing 2 commits to support PCH rootbus for Intel VMD enhancements.

Commit:
PCI: vmd: Clean up vmd_enable_domain function

This function is too long and needs to be shortened to make it more readable.
Certain VMD devices like 0xAD0B has two bus ranges, current implementation
supports only one range. This clean up is a prework for enablement additional
VMD bus range.

Signed-off-by: Szymon Durawa <szymon.durawa@linux.intel.com>
Suggested-by: Nirmal Patel <nirmal.patel@linux.intel.com>

Commit:
PCI: vmd: Add VMD PCH rootbus support

Starting from 0xAD0B VMD enhacement introduces separate rootbus for PCH.
It means that all 3 MMIO BARs exposed by VMD are shared now between CPU
IOC and PCH. This patch adds PCH bus enumeration and MMIO management
for devices with VMD enhancement support.

Signed-off-by: Szymon Durawa <szymon.durawa@linux.intel.com>
Suggested-by: Nirmal Patel <nirmal.patel@linux.intel.com>

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -74,14 +83,21 @@ enum vmd_features {
 	VMD_FEAT_BIOS_PM_QUIRK		= (1 << 5),
+
+	/*
+	 * Some client VMD devices have their VMD rootports connected to both
```
Nie wspominaj tutaj, że mówimy tu o "client" ponieważ nikogo to nie obchodzi. Część ludzi nie ogarnie nawet co masz na myśli. 
To co jest ważne to fakty. a fakt jest taki:

"Starting from Intel Lunar Lake, VMD driver might be connected to PCH rootbus" (jeżeli dobrze pamiętam, to Lunar lake).

sdurawa replied:
Fixed.

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -364,10 +382,26 @@ static void vmd_remove_irq_domain(struct vmd_dev *vmd)
+	 * PCI_VMD_PRIMARY_PCH_BUS but original value is 0xE1 which is stored
+	 * in vmd->busn_start_pch.
+	 */
+	if (vmd_has_pch_rootbus(vmd) && bus->number == PCI_VMD_PRIMARY_PCH_BUS)
```
wydaje mi się, że te warunki są jednoznaczne w tym miejscu i możesz sprawdzić tylko jeden z nich.

sdurawa replied:
Fixed.

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -364,10 +382,26 @@ static void vmd_remove_irq_domain(struct vmd_dev *vmd)
+	if (vmd_has_pch_rootbus(vmd) && bus->number == PCI_VMD_PRIMARY_PCH_BUS)
+		busnr_ecam = vmd->busn_start_pch - vmd->busn_start;
+	else
+		busnr_ecam = bus->number - vmd->busn_start;
```
jeżeli zachodzi koniecznośc takiego WA zdecydowanie czytelniejsze, będzie, jeżeli stworzysz pole (jeżeli dobrze rozumiem) `vmd->busn_offset[2]` i będziesz tam przypisywał wartość:
```
vmd->busn_offset[0] = bus->number;
vmd->busn_offset[1] = 0xE1;
```

 Chodzi mi o to że nazywanie pól pod pch ogranicza rozszerzalność tego rozwiązania (jeżeli ktoś za parę lat doda inny pcie bus ciężko będzie utrzymać odpowiednie nazewnictwo).

sdurawa replied:
Fixed.

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -132,13 +148,15 @@ struct vmd_dev {
+	struct resource		resources[6];
 	struct irq_domain	*irq_domain;
 	struct pci_bus		*bus;
+	struct pci_bus		*bus_pch;
```
nie da rady zmergować busów w tablice?

sdurawa replied:
dodac define na ilosc busow

sdurawa replied:
Fixed.

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -767,17 +814,253 @@ static int vmd_pm_enable_quirk(struct pci_dev *pdev, void *userdata)
+		};
+
+		/*
+		 * Update first cfgbar range (IOC) with value stored
```
nie musisz tłumaczyć kodu, każdy ogarnięty developer widzi co robisz. To co nas interesuje to dlaczego jest taka potrzeba. Najelpszy byłby link do specyfikacji jeżeli jest to PCIe spek. Pewnie nie jest.

sdurawa replied:
Fixed.

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -767,17 +814,253 @@ static int vmd_pm_enable_quirk(struct pci_dev *pdev, void *userdata)
 	struct resource *res;
 	u32 upper_bits;
 	unsigned long flags;
+	u32 vmd_enh_membar1_offset = 0;
```
myślę że `vmd_membar1_pch_offset` bedzie dokładniejsze

sdurawa replied:
Fixed.

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -767,17 +814,253 @@ static int vmd_pm_enable_quirk(struct pci_dev *pdev, void *userdata)
+		/*
+		 * Update first MEMBAR1 range (IOC), memory start does not
+		 * change, memory end is one less than the offset stored
+		 * in PCI_MEMBAR1_OFFSET.
```
podobnie tutaj, widzę co robisz. Muszę wiedzieć dlaczego to robisz.

sdurawa replied:
Fixed.

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -767,17 +814,253 @@ static int vmd_pm_enable_quirk(struct pci_dev *pdev, void *userdata)
+ *
+ * The only way we could use a 64-bit non-prefetchable MEMBAR is
+ * if its address is <4GB so that we can convert it to a 32-bit
+ * resource.  To be visible to the host OS, all VMD endpoints must
```
Double space.

sdurawa replied:
Fixed.

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -767,17 +814,253 @@ static int vmd_pm_enable_quirk(struct pci_dev *pdev, void *userdata)
+}
+
+/*
+ * Align resource information the same as for vmd_configure_membar1
```
Ten komentarz nic nie wnosi.

sdurawa replied:
removed.

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -767,17 +814,253 @@ static int vmd_pm_enable_quirk(struct pci_dev *pdev, void *userdata)
+ * Align resource information the same as for vmd_configure_membar1
+ * function.
+ */
+static void vmd_configure_membar2(struct vmd_dev *vmd,
```
te dwie funkcje `vmd_configure_membar1` i `vmd_configure_membar2` wydaje mi się, że można połączyć w jedną ponieważ kod jest niemal indentyczny (poza resorces, które updatujesz) oczywiście są one sobie odpowiadające więc jesteś w stanie odnosić się po indexach.

mtkaczyk replied:
no i wtedy możesz ładnie opisać co się znajduje w poszczególnych resources (jeżeli oczywiscie merge tych funkcji jest możliwy).

sdurawa replied:
Fixed.

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -767,17 +814,253 @@ static int vmd_pm_enable_quirk(struct pci_dev *pdev, void *userdata)
+	struct resource *res;
+	u32 upper_bits, reg;
+	unsigned long flags;
+	u64 vmd_enh_membar2_offset = 0;
```
podobnie jak wyżej, proszę zmień `vmd_membar2_pch_offset`

sdurawa replied:
Fixed.

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -767,17 +814,253 @@ static int vmd_pm_enable_quirk(struct pci_dev *pdev, void *userdata)
+	}
+
+	/*
+	 * VMD enhacement specific: pci_scan_bridge_extend() in probe.c
```
skróciłbym to trochę, i na razie podkreślił że to jest te miejsce przez które robimy RFC
```
This is a workaround for pci_scan_bridge_extend() code.
It assigns setup as broken when primary != bus->number.
For PCH rootbus primary is not "hard-wired to 0".
To avoid this, vmd->bus_pch->number and vmd->bus_pch->primary  are updated to same value.
```

sdurawa replied:
Fixed.

sdurawa commented on drivers/pci/controller/vmd.c:
```diff
@@ -31,6 +31,15 @@
 #define PCI_REG_VMLOCK		0x70
 #define MB2_SHADOW_EN(vmlock)	(vmlock & 0x2)
 
```
VMD enhacement nie zmienia prezentacji urzadzen

Commit:
PCI: vmd: Add VMD PCH rootbus support

Starting from 0xAD0B VMD enhacement introduces separate rootbus for PCH.
It means that all 3 MMIO BARs exposed by VMD are shared now between CPU
IOC and PCH. This patch adds PCH bus enumeration and MMIO management
for devices with VMD enhancement support.

Signed-off-by: Szymon Durawa <szymon.durawa@linux.intel.com>
Suggested-by: Nirmal Patel <nirmal.patel@linux.intel.com>

Commit:
PCI: vmd: Clean up vmd_enable_domain function

This function is too long and needs to be shortened to make it more readable.
Certain VMD devices like 0xAD0B has two bus ranges, current implementation
supports only one range. This clean up is a prework for enablement additional
VMD bus range.

Signed-off-by: Szymon Durawa <szymon.durawa@linux.intel.com>
Suggested-by: Nirmal Patel <nirmal.patel@linux.intel.com>

Commit:
Starting from Intel Arrow Lake VMD enhacement introduces separate rootbus for PCH.
It means that all 3 MMIO BARs exposed by VMD are shared now between CPU
IOC and PCH. This patch adds PCH bus enumeration and MMIO management
for devices with VMD enhancement support.

Signed-off-by: Szymon Durawa <szymon.durawa@linux.intel.com>
Suggested-by: Nirmal Patel <nirmal.patel@linux.intel.com>

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -23,6 +23,8 @@
 #define VMD_MEMBAR1	2
 #define VMD_MEMBAR2	4
 
+#define VMD_ROOTBUS_SIZE 2
```
imo `VMD_ROOTBUS_COUNT`

sdurawa replied:
Fixed.

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -132,13 +164,14 @@ struct vmd_dev {
 	struct irq_domain	*irq_domain;
-	struct pci_bus		*bus;
-	u8			busn_start;
+	struct pci_bus		*bus[VMD_ROOTBUS_SIZE];
```
you can eventually consider to group it into struct but it is fine for me.

sdurawa replied:
Fixed.

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -31,9 +33,32 @@
+};
+
+enum vmd_rootbus {
+	VMD_ROOTBUS0,
```
imo `VMD_ROOTBUS_0`, but is is fine anyway

sdurawa replied:
Fixed.

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -364,10 +397,27 @@ static void vmd_remove_irq_domain(struct vmd_dev *vmd)
+	 * PCI_VMD_PRIMARY_PCH_BUS but original value is 0xE1 which is stored
+	 * in vmd->busn_start[VMD_ROOTBUS1].
+	 */
+	if (vmd_has_pch_rootbus(vmd) && bus->number == PCI_VMD_PRIMARY_PCH_BUS)
```
you can detect it by:
`if (&(vmd->bus[VMD_ROOTBUS1]) == bus)`

and I think it is better readable, because you are not reaying on hack but on adress.
Eventually, you can consider passing `enum vmd_rootbus` instead of bus, and deference it here.

If you passing `struct vmd_dev *vmd` and passing `struct pci_bus *bus` placed inside `vmd` the struct is not mandatory because you can obtain it easy.

sdurawa replied:
Condition is rewritten using ? operand.

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -364,10 +397,27 @@ static void vmd_remove_irq_domain(struct vmd_dev *vmd)
+	unsigned char bus_number = 0;
+
+	/*
+	 * VMD enhacement specific: for PCH rootbus, bus number is set to
```
Please remove `VMD enhancement specyfic`. It has nothing to do with VMD and it is not specific, it is a hack for kernel code elsewhere.


Please add reference to `vmd_create_pch_bus`. Simple
```
See comment in vmd_create_pch_bus().
```
Should be fine.

sdurawa replied:
Fixed

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -644,13 +694,39 @@ static int vmd_get_bus_number_start(struct vmd_dev *vmd)
+			vmd->busn_start[VMD_ROOTBUS0] = 224;
+			break;
+		case 3:
+			if (features & VMD_FEAT_HAS_PCH_ROOTBUS) {
```
I would revert the logic but it is my personal preference:
```c
if (!(features & VMD_FEAT_HAS_PCH_ROOTBUS)) {
	pci_err(dev, "Unknown Bus Offset Setting (%d)\n", BUS_RESTRICT_CFG(reg));
	return -ENODEV;
}

/* IOC start bus */
vmd->busn_start[VMD_ROOTBUS0] = 224;
/* PCH start bus */
vmd->busn_start[VMD_ROOTBUS1] = 225;
```

sdurawa replied:
Fixed.

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -644,13 +694,39 @@ static int vmd_get_bus_number_start(struct vmd_dev *vmd)
+				return -ENODEV;
+			}
+			break;
+		case 3:
```
unnecessary doubled statement, probably typo :)

sdurawa replied:
Fixed.

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -644,13 +694,39 @@ static int vmd_get_bus_number_start(struct vmd_dev *vmd)
+				vmd->busn_start[VMD_ROOTBUS1] = 225;
+			} else {
+				pci_err(dev,
+					"Unknown Bus Offset Setting (%d)\n",
```
`"Unknown Bus Offset Setting (%d)\n"` is not really helpful.

What about:
`Determined 3 vmd configs but VMD in PCH bus is not supported, aborting.`

Probably you can find better wording, I didn't check it.

sdurawa replied:
Fixed.

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -767,17 +843,226 @@ static int vmd_pm_enable_quirk(struct pci_dev *pdev, void *userdata)
+}
+
+static void vmd_configure_membar(struct vmd_dev *vmd,
+				 u8 membar_number, u8 resource_number,
```
`u8 membar_number` should be enum I think.

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -767,17 +843,226 @@ static int vmd_pm_enable_quirk(struct pci_dev *pdev, void *userdata)
+	}
+}
+
+static void vmd_configure_membar(struct vmd_dev *vmd,
```
Description upon the function would be recommended and helpful.
https://www.kernel.org/doc/html/latest/doc-guide/kernel-doc.html

sdurawa replied:
Description added.

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -767,17 +843,226 @@ static int vmd_pm_enable_quirk(struct pci_dev *pdev, void *userdata)
+		 */
+		vmd_configure_membar(vmd, VMD_MEMBAR1, VMD_RESOURCE_MEMBAR1, 0,
+				     pch_membar1_offset, NULL);
+		vmd_configure_membar(vmd, VMD_MEMBAR2, VMD_RESOURCE_MEMBAR2,
```
What I don' like here is repeated initialization please consider reverting order. or use if/else

sdurawa replied:
Fixed.

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -644,13 +694,39 @@ static int vmd_get_bus_number_start(struct vmd_dev *vmd)
+				pci_err(dev,
+					"Unknown Bus Offset Setting (%d)\n",
+					BUS_RESTRICT_CFG(reg));
+				return -ENODEV;
```
Please consider other error codes (in each case now -ENODEV is returned from various functions)

sdurawa replied:
Fixed.

Commit:
Source code modifications according to review remarks.

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -31,9 +31,34 @@
 #define MB2_SHADOW_EN(vmlock)	(vmlock & 0x2)
 
+#define PCI_VMD_PRIMARY_PCH_BUS 0x80
+#define PCI_REG_BUSRANGE0 0xC8
```
Probably we should make all defines starting with `PCI_VMD` or just `VMD`. We should avoid name confusion with different components.

If you would like to stay with `PCI_*` then it is not right place for those defines because they are probably global and should be placed in include/linux/pci_regs.h - we don't want that.

sdurawa replied:
Fixed.

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -31,9 +31,34 @@
 #define MB2_SHADOW_SIZE		16
 
+enum vmd_resource {
+	VMD_RESOURCE_CFGBAR,
```
`VMD_RESOURCE_CFGBAR = 0` to force initialization from 0

```suggestion
VMD_RESOURCE_CFGBAR = 0,
```

sdurawa replied:
Fixed.

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -74,14 +99,21 @@ enum vmd_features {
 				 VMD_FEAT_HAS_BUS_RESTRICTIONS |	\
 				 VMD_FEAT_OFFSET_FIRST_VECTOR |		\
-				 VMD_FEAT_BIOS_PM_QUIRK)
+				 VMD_FEAT_BIOS_PM_QUIRK |	\
```
Looks like to alligned same way as above but it could be presentation issue:
```suggestion
				 VMD_FEAT_BIOS_PM_QUIRK |		\
```

sdurawa replied:
Fixed.

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -31,9 +31,34 @@
+};
+
+enum vmd_rootbus {
+	VMD_ROOTBUS_0,
```
```suggestion
	VMD_ROOTBUS_0 = 0,
```

sdurawa replied:
Fixed.

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -132,10 +164,10 @@ struct vmd_dev {
-	struct pci_bus		*bus;
-	u8			busn_start;
+	struct pci_bus		*bus[VMD_ROOTBUS_COUNT];
+	u8			busn_start[VMD_ROOTBUS_COUNT];
```
It is asking to be placed in structure.

```
struct vmd_pci_bus {
    struct pci_bus *bus;
    u8 busn_start;
```

```suggestion
struct vmd_pci_bus vmd_bus[VMD_ROOTBUS_COUNT];
```
but I think it is fine either way

sdurawa replied:
No changes here.

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -892,70 +1148,49 @@ static int vmd_enable_domain(struct vmd_dev *vmd, unsigned long features)
-	pci_scan_child_bus(vmd->bus);
-	vmd_domain_reset(vmd);
+	WARN(sysfs_create_link(&vmd->dev->dev.kobj,
+			       &vmd->bus[VMD_ROOTBUS_0]->dev.kobj, "domain"),
```
Shouldn't we create domain link for pch attached ROOTBUS_1?

sdurawa replied:
No, there is only required for VMD domain. When creating sys link for ROOTBUS 1 there is an error "cannot create duplicate filename '/devices/pci0000:00/0000:00:0e.0/domain'".

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -644,13 +693,28 @@ static int vmd_get_bus_number_start(struct vmd_dev *vmd)
+				vmd->busn_start[VMD_ROOTBUS_1] = 225;
+			} else {
+				pci_err(dev,
+					"VMD Bus Restriction detected type %d,",
```
```suggestion
				pci_err(dev, "VMD Bus Restriction detected type %d, but PCH Rootbus is not supported, aborting.",
					   BUS_RESTRICT_CFG(reg));

```

mtkaczyk replied:
This messages cannot be divided because kernel is not guarantying that they will be flushed together, one by one.

sdurawa replied:
Fixed.

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -767,17 +831,239 @@ static int vmd_pm_enable_quirk(struct pci_dev *pdev, void *userdata)
+	if (!upper_bits)
+		flags &= ~IORESOURCE_MEM_64;
+
+	snprintf(name, sizeof(name), "VMD MEMBAR%d", membar_number/2);
```
`strncat` should be avoided :)

```suggestion
	snprintf(name, sizeof(name), "VMD MEMBAR%d %s", membar_number / 2,
	         resource_number > VMD_RESOURCE_MEMBAR_2 : "PCH", "");
```

sdurawa replied:
Fixed with resource_number > VMD_RESOURCE_MEMBAR_2 ? "PCH":""

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -767,17 +831,239 @@ static int vmd_pm_enable_quirk(struct pci_dev *pdev, void *userdata)
+ */
+static void vmd_configure_membar(struct vmd_dev *vmd,
+				 enum vmd_resource resource_number,
+				 u8 membar_number, resource_size_t start_offset,
```
``` suggestion
 				enum vmd_resource membar_number, resource_size_t start_offset,
```

sdurawa replied:
membar number  does not have an enum, it is a standalone define.

mtkaczyk commented on drivers/pci/controller/vmd.c:
```diff
@@ -767,17 +831,239 @@ static int vmd_pm_enable_quirk(struct pci_dev *pdev, void *userdata)
+
+	if (!vmd->bus[VMD_ROOTBUS_1]) {
+		pci_free_resource_list(&resources_pch);
+		vmd_remove_irq_domain(vmd);
```
Should we? We didn't register PCH MSI yet.

mtkaczyk replied:
Shoudn't we call `pci_stop_root_bus` and `pci_remove_root_bus`?

sdurawa replied:
correct, irq domain cannot be removed here, I have added stopping and removig PCH bus instead.

mtkaczyk reviewed (commented):
+ Please update module version
+ Please update patches author (currently it is `sdurawa`)

Commit:
Replace PCI with VMD prefix for variables, align code indentation.
